internal.sh

```diff
@@ -0,0 +1,86 @@
+            exit 1
+        elif [ ! -f ${OPENSHIFT_PULL_SECRET_PATH} ]; then
+            echo "Provided OPENSHIFT_PULL_SECRET_PATH (${OPENSHIFT_PULL_SECRET_PATH}) does not exists"
+            exit 1
+        fi
+}
+
+run_preflight_checks
+
+MIRROR=${MIRROR:-https://mirror.openshift.com/pub/openshift-v4/x86_64/clients/ocp}
+
+# If user defined the OPENSHIFT_VERSION environment variable then use it.
+if test -n "${OPENSHIFT_VERSION-}"; then
+    OPENSHIFT_RELEASE_VERSION=${OPENSHIFT_VERSION}
+    echo "Using release ${OPENSHIFT_RELEASE_VERSION} from OPENSHIFT_VERSION"
+else
+    OPENSHIFT_RELEASE_VERSION="$(curl -L "${MIRROR}"/candidate-4.11/release.txt | sed -n 's/^ *Version: *//p')"
+    if test -n "${OPENSHIFT_RELEASE_VERSION}"; then
+        echo "Using release ${OPENSHIFT_RELEASE_VERSION} from the latest mirror"
+    else
+        echo "Unable to determine an OpenShift release version.  You may want to set the OPENSHIFT_VERSION environment variable explicitly."
+        exit 1
+    fi
+fi
+
+if test -z "${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE-}"; then
+    OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="$(curl -L "${MIRROR}/${OPENSHIFT_RELEASE_VERSION}/release.txt" | sed -n 's/^Pull From: //p')"
+elif test -n "${OPENSHIFT_VERSION-}"; then
+    echo "Both OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE and OPENSHIFT_VERSION are set, OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE will take precedence"
+    echo "OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: $OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE"
+    echo "OPENSHIFT_VERSION: $OPENSHIFT_VERSION"
+fi
+echo "Setting OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE to ${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE}"
+
+mkdir -p openshift-clients/linux
+curl -L "${MIRROR}/${OPENSHIFT_RELEASE_VERSION}/openshift-client-linux-${OPENSHIFT_RELEASE_VERSION}.tar.gz" | tar -zx -C openshift-clients/linux oc
+OC=./openshift-clients/linux/oc
+
+function patch_and_push_image() {
+    local image_name=$1
+    image=$(${OC} adm release info -a ${OPENSHIFT_PULL_SECRET_PATH} ${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE} --image-for=${image_name})
+    vcs_ref=$(${OC} image info -a ${OPENSHIFT_PULL_SECRET_PATH} ${image} -ojson | jq -r '.config.config.Labels."vcs-ref"')
+    url=$(${OC} image info -a ${OPENSHIFT_PULL_SECRET_PATH} ${image} -ojson | jq -r '.config.config.Labels.url')
+    url_tag=$(basename ${url})
+    rhpkg clone containers/crc-${image_name}
+    pushd crc-${image_name}
+    git remote add upstream git://pkgs.devel.redhat.com/containers/ose-${image_name}
+    git fetch --all
+    git checkout --track  origin/rhaos-4.11-rhel-8
+    git merge ${vcs_ref}
+    git push -f origin HEAD
+    rhpkg container-build  --target crc-1-rhel-8-candidate
+    popd
+    skopeo copy --dest-authfile ${OPENSHIFT_PULL_SECRET_PATH} --all docker://registry-proxy.engineering.redhat.com/rh-osbs/openshift-crc-${image_name}:${url_tag} docker://quay.io/crcont/openshift-crc-${image_name}:${vcs_ref}
```

Comment:
Do we want to tag the image with`vcs_ref`? Seems like openshift version would be easier to use for our purpose?

Reply:
This is for safer side and easy to track the image build from downstream dist-git commit (just looking at it) then inspecting the image.

Reply:
"safer side"? What is safer? vcs-ref works, but either openshift version or upstream image digest look like more direct ways of addressing the image, without a lot of drawbacks. Are we going to want to often do map the images we build to ose-* dist-git commits?

Reply:
> Are we going to want to often do map the images we build to ose-* dist-git commits?

We do need to for each bundle creation side (not sure what would the frequency for it).

`version` for each commit is same `4.11.0` and also upstream commit.id can be same for some of the downstream commits. 
```
commit b4a2cedf36233fb3ca12b614ff94d1be266d1f13 (tag: v4.11.0-202209051237.p0.g74584b7.assembly.stream, upstream/rhaos-4.11-rhel-8)
Author: AOS Automation Release Team <noreply@redhat.com>
Date:   Mon Sep 5 12:46:08 2022 +0000

    # https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31762/
    MaxFileSize: 104857600
    io.openshift.build.commit.id: 74584b7e3c7fd13b71bcb93c517d070d54a29409
    io.openshift.build.source-location: https://github.com/openshift/cluster-kube-controller-manager-operator
    jenkins.url: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31762/
    release: 202209051237.p0.g74584b7.assembly.stream
    version: v4.11.0

commit e5e27284b097c44f05fcf0eeffa7355460405f49 (tag: v4.11.0-202209012047.p0.g7d99298.assembly.stream)
Author: AOS Automation Release Team <noreply@redhat.com>
Date:   Thu Sep 1 20:55:50 2022 +0000

    # https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31618/
    MaxFileSize: 104857600
    io.openshift.build.commit.id: 7d99298a2cabdea1b26edb231a0c9719ffcee817
    io.openshift.build.source-location: https://github.com/openshift/cluster-kube-controller-manager-operator
    jenkins.url: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31618/
    release: 202209012047.p0.g7d99298.assembly.stream
    version: v4.11.0

commit b5985c4178f1588436b98079d96b2a943530a6ab (tag: v4.11.0-202208291725.p0.g3b7135b.assembly.stream)
Author: AOS Automation Release Team <noreply@redhat.com>
Date:   Mon Aug 29 17:46:09 2022 +0000

    # https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31511/
    MaxFileSize: 104857600
    io.openshift.build.commit.id: 3b7135b462320afc892a2e031c47298408dbef9f
    io.openshift.build.source-location: https://github.com/openshift/cluster-kube-controller-manager-operator
    jenkins.url: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31511/
    release: 202208291725.p0.g3b7135b.assembly.stream
    version: v4.11.0
commit e1399cd79ad75e612787d4893976bc15187263fd (tag: v4.11.0-202208250725.p0.g3b7135b.assembly.stream)
Author: AOS Automation Release Team <noreply@redhat.com>
Date:   Thu Aug 25 07:46:08 2022 +0000

    # https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31404/
    MaxFileSize: 104857600
    io.openshift.build.commit.id: 3b7135b462320afc892a2e031c47298408dbef9f
    io.openshift.build.source-location: https://github.com/openshift/cluster-kube-controller-manager-operator
    jenkins.url: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31404/
    release: 202208250725.p0.g3b7135b.assembly.stream
    version: v4.11.0

commit 9670a96da9ebfb9091dc06031e9ef8d4bfbd57ae (tag: v4.11.0-202208231635.p0.g3b7135b.assembly.stream)
Author: AOS Automation Release Team <noreply@redhat.com>
Date:   Tue Aug 23 16:50:56 2022 +0000

    # https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31338/
    MaxFileSize: 104857600
    io.openshift.build.commit.id: 3b7135b462320afc892a2e031c47298408dbef9f
    io.openshift.build.source-location: https://github.com/openshift/cluster-kube-controller-manager-operator
    jenkins.url: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31338/
    release: 202208231635.p0.g3b7135b.assembly.stream
    version: v4.11.0

commit 4a19080078ac3744b4537bed5157944bf477e9fb (tag: v4.11.0-202208190905.p0.g3b7135b.assembly.stream)
Author: AOS Automation Release Team <noreply@redhat.com>
Date:   Fri Aug 19 09:15:44 2022 +0000

    # https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31249/
    MaxFileSize: 104857600
    io.openshift.build.commit.id: 3b7135b462320afc892a2e031c47298408dbef9f
    io.openshift.build.source-location: https://github.com/openshift/cluster-kube-controller-manager-operator
    jenkins.url: https://saml.buildvm.hosts.prod.psi.bos.redhat.com:8888/job/aos-cd-builds/job/build%252Focp4/31249/
    release: 202208190905.p0.g3b7135b.assembly.stream
    version: v4.11.0

```

Reply:
@cfergeau what you suggested to make easy to detect stuff for upstream side and then go downstream, what's current solution is downstream => upstream. I implemented it and now image have openshift version tags. check https://quay.io/repository/crcont/openshift-crc-cluster-kube-apiserver-operator?tab=tags